### ISSUE TITLE
User login through Dex OIDC and gets "RBAC: access denied" web page several minutes after everything looks fine 

### ISSUE THREAD

**Issue by joey100** [kubeflow/kubeflow#5077]
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

Deploy Kubeflow with KFDef kfctl_istio_dex.v1.0.2

Login to the Kubeflow dashboard through Dex OIDC and pages show normal, such as Home page/Pipelines page/Katib page/Notebooks Servers page.

Several minutes later( probably 3 or 5 minutes), refresh the page, it shows "RBAC: access denied" information. Navigate to other pages, still show "RBAC: access denied" information. From Chrome developer tool we see 403 error.

Several minutes later, refresh the page, it shows normal again.  Then several minutes later, refresh the page, it shows "RBAC: access denied" again.  It's like a cycle of "good" -- "bad" -- "good" -- ...

**What did you expect to happen:**
Login to the Kubeflow dashboard through Dex OIDC and pages show normal always, but not just several minutes.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
Once we face the "RBAC: access denied" issue, we check authservice-0  and dex pod logs, not much useful information, even no error logs. If we delete the authservice-0 pod and let a new pod run, then refresh the page it will show the login page or the normal page. 

It seems not a http cache control issue, right? I don't know where else to check, the ingressgateway logs /envoyfilter config look good. And authservice-0 "SESSION_MAX_AGE" uses "86400" as default.

By the way, we use nginx proxy as the frontend loadbalancer, to proxy the request to backend ingressgateway svc.

**Environment:**

- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard): 1.0.x    Build: v1beta1 | Dashboard: v0.0.2 | Isolation-Mode: multi-user
- kfctl version: (use `kfctl version`):
- Kubernetes platform: (e.g. `minikube`) GKE ( we use dex, not IAP, so the GKE is like a on-prem cluster)
- Kubernetes version: (use `kubectl version`): Client: 1.15.1; Server: 1.15.11
- OS (e.g. from `/etc/os-release`):

**Comment by joey100** [comment]
@yanniszark @jlewi @pinglin @vincentarthur

**Comment by jlewi** [comment]
@joey100 You will need to ask someone who is familiar with the DEX support. DEX is really at the boundary of KF so you might have better luck in the DEX forums. I believe the DEX and ISTIO solution in KF does have some KF specific parts (e.g. using ISTIO envoy filters to redirect you through the login).

**Comment by joey100** [comment]
> @joey100 You will need to ask someone who is familiar with the DEX support. DEX is really at the boundary of KF so you might have better luck in the DEX forums. I believe the DEX and ISTIO solution in KF does have some KF specific parts (e.g. using ISTIO envoy filters to redirect you through the login).

@jlewi  Thanks a lot for your quick response! However, according to below doc: https://journal.arrikto.com/kubeflow-authentication-with-istio-dex-5eafdfac4782
and the pic below, this flow shows like the 403 error was returned by authservice, it doesn't go to Dex yet. Correct me if I am wrong.

![image](https://user-images.githubusercontent.com/18437960/85104830-e7cce180-b23b-11ea-8d7d-25fa2e1af640.png)

@yanniszark  Could you please also take a look at the authservice/istio, and advise us, thanks a lot!

**Comment by yanniszark** [comment]
Hi @joey100! Error 403 could be coming from the authservice indeed, but it's not certain.
I would start by getting logs from the authservice and looking for clues there.
In addition, if you have made any change from the default deployment, it would be good to know.

**Comment by joey100** [comment]
> Hi @joey100! Error 403 could be coming from the authservice indeed, but it's not certain.
> I would start by getting logs from the authservice and looking for clues there.
> In addition, if you have made any change from the default deployment, it would be good to know.

@yanniszark  Thanks a lot! What is so strange here is that no error logs from authservice/dex/ingressgateway pods when the issue happens. And I don't do anything, several minutes later, the web page shows normal again! After several minutes, the web page shows "RBAC: access denied" again! Still no error logs!  Is there some timeout settings at somewhere we dont know to control this 'behavior' ?  I'll put the pods logs here.

**Comment by joey100** [comment]
Authservice pod log:
![image](https://user-images.githubusercontent.com/18437960/85124609-c4fef500-b25c-11ea-9017-f85217b2ec41.png)


Dex pod log:
![image](https://user-images.githubusercontent.com/18437960/85124657-d942f200-b25c-11ea-80af-9f136e046699.png)

**Comment by vincentarthur** [comment]
Hi @yanniszark , do you think authservice can be changed to service or deployment instead of Stateful set, due to scalability perspective ?

**Comment by yanniszark** [comment]
@joey100 if you have another proxy in front of the Istio Gateway (you mentioned NGINX), can you remove it and see if the problem persists? It may be a caching issue. We haven't seen this problem in any other deployment so far.

> Hi @yanniszark , do you think authservice can be changed to service or deployment instead of Stateful set, due to scalability perspective ?

Hi @vincentarthur, scalability of the authservice is currently bottlenecked by having a local database for the sessions. Feel free to open a separate issue for this!

**Comment by vincentarthur** [comment]
> @joey100 if you have another proxy in front of the Istio Gateway (you mentioned NGINX), can you remove it and see if the problem persists? It may be a caching issue. We haven't seen this problem in any other deployment so far.

We tried on another environment without proxy, it also hit same problem. Also we tried the latest docker image of `gcr.io/arrikto/kubeflow/oidc-authservice`, does not work either. Some parameters changed..

As we are in the middle and blocked by this case. Any chance to reproduce at your end and investigate as well ? (we are also doing same thing..)

**Comment by vincentarthur** [comment]
Looks like the `rbac: access denied` was due to Envoy stream idle timeout, which disconnect all connections every 5 mins (default value). But still seeking proper way to configure Envoy `stream_idle_timeout` in Istio level.

**Comment by vincentarthur** [comment]
Eventually - we found the solution as below -

- install Kubeflow 1.0.2 via `kfctl_k8s_istio.yaml`
- Add `volumeClaimTemplates: []` to `kustomize/oidc-authservice/base/statefulset.yaml` (output from kfctl_k8s_dex.yaml)
- manually install `dex` and `oidc-authservice` to cluster (ride on KF1.0.2)

Then access the UI, the `rbac: access denied` issue is gone.

So - i think the issue should be related to Istio version.

/close

**Comment by vincentarthur** [comment]
@joey100  - please reopen. A mistake made before, when using kfctl_k8s_istio.yaml, by default, `rbacClusterConfig` is set `OFF` so that `rbac:access denied` not shown...

Issue remained...
